File: agent/stats.ts
import { spawn } from "node:child_process";
import os from "node:os";

/**
 * Self-contained port of the main app's `SystemStatsService` : not imported
 * from there since the agent is a separately-deployed binary with no access
 * to the main app's source tree at runtime. Keep the two in sync by hand if
 * one changes; they're small and unlikely to drift silently.
 */
export interface SystemStats {
	cpuPercent: number;
	memTotalMb: number;
	memUsedMb: number;
	memPercent: number;
	diskTotalMb: number | null;
	diskUsedMb: number | null;
	diskPercent: number | null;
	gpu: {
		name: string;
		memTotalMb: number;
		memUsedMb: number;
		utilizationPercent: number;
	} | null;
}

interface CpuSample {
	idle: number;
	total: number;
}

let lastCpuSample: CpuSample | null = null;

function sampleCpu(): CpuSample {
	let idle = 0;
	let total = 0;
	for (const cpu of os.cpus()) {
		idle += cpu.times.idle;
		total +=
			cpu.times.user +
			cpu.times.nice +
			cpu.times.sys +
			cpu.times.idle +
			cpu.times.irq;
	}
	return { idle, total };
}

function run(cmd: string, args: string[]): Promise<string | null> {
	return new Promise((resolve) => {
		try {
			const child = spawn(cmd, args);
			let out = "";
			child.stdout.on("data", (d) => {
				out += d.toString();
			});
			child.on("error", () => resolve(null));
			child.on("close", (code) => resolve(code === 0 ? out : null));
		} catch {
			resolve(null);
		}
	});
}

async function diskStats(): Promise<{
	totalMb: number;
	usedMb: number;
	percent: number;
} | null> {
	const out = await run("df", ["-Pk", "."]);
	if (!out) {
		return null;
	}
	const line = out.trim().split("\n").at(-1);
	const parts = line?.trim().split(/\s+/);
	if (!parts || parts.length < 5) {
		return null;
	}
	const totalKb = Number.parseInt(parts[1], 10);
	const usedKb = Number.parseInt(parts[2], 10);
	if (!(Number.isFinite(totalKb) && Number.isFinite(usedKb)) || totalKb === 0) {
		return null;
	}
	return {
		percent: Math.round((usedKb / totalKb) * 100),
		totalMb: Math.round(totalKb / 1024),
		usedMb: Math.round(usedKb / 1024),
	};
}

async function gpuStats(): Promise<SystemStats["gpu"]> {
	const out = await run("nvidia-smi", [
		"--query-gpu=name,memory.total,memory.used,utilization.gpu",
		"--format=csv,noheader,nounits",
	]);
	if (!out) {
		return null;
	}
	const [name, memTotal, memUsed, util] = out
		.trim()
		.split(",")
		.map((s) => s.trim());
	if (!name) {
		return null;
	}
	return {
		memTotalMb: Number.parseInt(memTotal, 10) || 0,
		memUsedMb: Number.parseInt(memUsed, 10) || 0,
		name,
		utilizationPercent: Number.parseInt(util, 10) || 0,
	};
}

export async function getSystemStats(): Promise<SystemStats> {
	const sample = sampleCpu();
	let cpuPercent = 0;
	if (lastCpuSample) {
		const idleDelta = sample.idle - lastCpuSample.idle;
		const totalDelta = sample.total - lastCpuSample.total;
		cpuPercent =
			totalDelta > 0 ? Math.round(100 * (1 - idleDelta / totalDelta)) : 0;
	}
	lastCpuSample = sample;

	const memTotalMb = Math.round(os.totalmem() / 1024 / 1024);
	const memFreeMb = Math.round(os.freemem() / 1024 / 1024);
	const memUsedMb = memTotalMb - memFreeMb;

	const [disk, gpu] = await Promise.all([diskStats(), gpuStats()]);

	return {
		cpuPercent: Math.min(100, Math.max(0, cpuPercent)),
		diskPercent: disk?.percent ?? null,
		diskTotalMb: disk?.totalMb ?? null,
		diskUsedMb: disk?.usedMb ?? null,
		gpu,
		memPercent: memTotalMb > 0 ? Math.round((memUsedMb / memTotalMb) * 100) : 0,
		memTotalMb,
		memUsedMb,
	};
}

File: internal/agent/stats.go
package agent

import (
	"bufio"
	"math"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"sync"
)

// GPUStats is the first NVIDIA GPU's reading.
type GPUStats struct {
	MemTotalMb         int    `json:"memTotalMb"`
	MemUsedMb          int    `json:"memUsedMb"`
	Name               string `json:"name"`
	UtilizationPercent int    `json:"utilizationPercent"`
}

// SystemStats is GET /v1/stats's answer, the same shape as the main app's own
// host stats so the dashboard renders a remote host exactly like the local one.
type SystemStats struct {
	CPUPercent  int       `json:"cpuPercent"`
	DiskPercent *int      `json:"diskPercent"`
	DiskTotalMb *int      `json:"diskTotalMb"`
	DiskUsedMb  *int      `json:"diskUsedMb"`
	GPU         *GPUStats `json:"gpu"`
	MemPercent  int       `json:"memPercent"`
	MemTotalMb  int       `json:"memTotalMb"`
	MemUsedMb   int       `json:"memUsedMb"`
}

type cpuSample struct {
	idle  uint64
	total uint64
}

// StatsSampler samples host usage. CPU% is a delta against the previous
// sample, so the sampler keeps the last one; the first call reports 0.
type StatsSampler struct {
	mu         sync.Mutex
	last       *cpuSample
	procRoot   string
	runCommand func(name string, args ...string) (string, bool)
}

// NewStatsSampler builds a sampler reading the real /proc and running the real
// df and nvidia-smi.
func NewStatsSampler() *StatsSampler {
	return &StatsSampler{procRoot: "/proc", runCommand: runCommand}
}

func runCommand(name string, args ...string) (string, bool) {
	output, err := exec.Command(name, args...).Output()
	if err != nil {
		return "", false
	}
	return string(output), true
}

// Sample reads CPU, memory, disk and GPU usage. Every call replaces the stored
// CPU sample. On a host without /proc (a macOS dev machine), CPU and memory
// read as zero rather than failing.
func (s *StatsSampler) Sample() SystemStats {
	stats := SystemStats{}

	if sample, ok := s.sampleCPU(); ok {
		s.mu.Lock()
		if s.last != nil && sample.total > s.last.total {
			idle := float64(sample.idle - s.last.idle)
			total := float64(sample.total - s.last.total)
			stats.CPUPercent = clamp(int(math.Round(100*(1-idle/total))), 0, 100)
		}
		s.last = &sample
		s.mu.Unlock()
	}

	if total, available, ok := s.memory(); ok {
		stats.MemTotalMb = int(math.Round(float64(total) / 1024))
		stats.MemUsedMb = stats.MemTotalMb - int(math.Round(float64(available)/1024))
		if total > 0 {
			stats.MemPercent = int(math.Round(float64(total-available) / float64(total) * 100))
		}
	}

	if totalMb, usedMb, percent, ok := s.disk(); ok {
		stats.DiskTotalMb, stats.DiskUsedMb, stats.DiskPercent = &totalMb, &usedMb, &percent
	}
	stats.GPU = s.gpu()
	return stats
}

func clamp(value, low, high int) int {
	return max(low, min(high, value))
}

// sampleCPU reads the aggregate line of /proc/stat. Total is user, nice,
// system, idle and irq, the same five counters the main app sums.
func (s *StatsSampler) sampleCPU() (cpuSample, bool) {
	file, err := os.Open(s.procRoot + "/stat")
	if err != nil {
		return cpuSample{}, false
	}
	defer func() { _ = file.Close() }()
	scanner := bufio.NewScanner(file)
	if !scanner.Scan() {
		return cpuSample{}, false
	}
	fields := strings.Fields(scanner.Text())
	if len(fields) < 7 || fields[0] != "cpu" {
		return cpuSample{}, false
	}
	values := make([]uint64, 7)
	for index := 1; index < 7; index++ {
		parsed, err := strconv.ParseUint(fields[index], 10, 64)
		if err != nil {
			return cpuSample{}, false
		}
		values[index] = parsed
	}
	user, nice, system, idle, irq := values[1], values[2], values[3], values[4], values[6]
	return cpuSample{idle: idle, total: user + nice + system + idle + irq}, true
}

// memory reads MemTotal and MemAvailable from /proc/meminfo, in kB.
// MemAvailable, not MemFree: page cache the kernel would hand back on demand
// isn't memory in use.
func (s *StatsSampler) memory() (int, int, bool) {
	file, err := os.Open(s.procRoot + "/meminfo")
	if err != nil {
		return 0, 0, false
	}
	defer func() { _ = file.Close() }()
	values := map[string]int{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		if len(fields) < 2 {
			continue
		}
		key := strings.TrimSuffix(fields[0], ":")
		if key == "MemTotal" || key == "MemAvailable" {
			if parsed, err := strconv.Atoi(fields[1]); err == nil {
				values[key] = parsed
			}
		}
	}
	total, hasTotal := values["MemTotal"]
	available, hasAvailable := values["MemAvailable"]
	return total, available, hasTotal && hasAvailable
}

// disk reads usage of the filesystem holding the working directory via df.
func (s *StatsSampler) disk() (int, int, int, bool) {
	output, ok := s.runCommand("df", "-Pk", ".")
	if !ok {
		return 0, 0, 0, false
	}
	lines := strings.Split(strings.TrimSpace(output), "\n")
	fields := strings.Fields(lines[len(lines)-1])
	if len(fields) < 5 {
		return 0, 0, 0, false
	}
	totalKb, totalErr := strconv.Atoi(fields[1])
	usedKb, usedErr := strconv.Atoi(fields[2])
	if totalErr != nil || usedErr != nil || totalKb == 0 {
		return 0, 0, 0, false
	}
	percent := int(math.Round(float64(usedKb) / float64(totalKb) * 100))
	return int(math.Round(float64(totalKb) / 1024)), int(math.Round(float64(usedKb) / 1024)), percent, true
}

// gpu reads the first NVIDIA GPU's name, memory and utilisation via
// nvidia-smi, or nil when there's none.
func (s *StatsSampler) gpu() *GPUStats {
	output, ok := s.runCommand(
		"nvidia-smi",
		"--query-gpu=name,memory.total,memory.used,utilization.gpu",
		"--format=csv,noheader,nounits",
	)
	if !ok {
		return nil
	}
	first := strings.Split(strings.TrimSpace(output), "\n")[0]
	parts := strings.Split(first, ",")
	for index := range parts {
		parts[index] = strings.TrimSpace(parts[index])
	}
	if len(parts) < 4 || parts[0] == "" {
		return nil
	}
	atoi := func(value string) int {
		parsed, _ := strconv.Atoi(value)
		return parsed
	}
	return &GPUStats{
		MemTotalMb:         atoi(parts[1]),
		MemUsedMb:          atoi(parts[2]),
		Name:               parts[0],
		UtilizationPercent: atoi(parts[3]),
	}
}
